<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Generate SNPs on given samples. — make_snp • tekkamaki</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- katex math --><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous"><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.js" integrity="sha384-7zkQWkzuo3B5mTepMUcHkMB5jZaolc2xDwL6VFqjFALcbeS9Ggm/Yr2r3Dy4lfFg" crossorigin="anonymous"></script><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/contrib/auto-render.min.js" integrity="sha384-43gviWU0YVjaDtb/GhzOouOXtZMP/7XUzwPTstBeZFe/+rCMvRwr4yROQP43s0Xk" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script><script src="../katex-auto.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate SNPs on given samples. — make_snp"><meta name="description" content="place_mutations() generates a SNP data.frame by randomly placing a fixed
number of mutations on a given genealogy.
It means that all the sites are perfectly linked with each other.
make_snp_chromosome() simulates a SNP data.frame on a chromosome using
place_mutations() with recombination.
make_snp() calls it for each chromosome in parallel.
Use RNGkind(&quot;L'Ecuyer-CMRG&quot;) and set.seed() to get reproducible results.
The number of CPU cores used can be configured via mc.cores option."><meta property="og:description" content="place_mutations() generates a SNP data.frame by randomly placing a fixed
number of mutations on a given genealogy.
It means that all the sites are perfectly linked with each other.
make_snp_chromosome() simulates a SNP data.frame on a chromosome using
place_mutations() with recombination.
make_snp() calls it for each chromosome in parallel.
Use RNGkind(&quot;L'Ecuyer-CMRG&quot;) and set.seed() to get reproducible results.
The number of CPU cores used can be configured via mc.cores option."><meta property="og:image" content="https://heavywatal.github.io/tekkamaki/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tekkamaki</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/tekkamaki.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/kinship.html">Extract Kinship</a></li>
    <li><a class="dropdown-item" href="../articles/snp.html">Simulate SNP pattern</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/heavywatal/tekkamaki/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Generate SNPs on given samples.</h1>
      <small class="dont-index">Source: <a href="https://github.com/heavywatal/tekkamaki/blob/HEAD/R/snp.R" class="external-link"><code>R/snp.R</code></a></small>
      <div class="d-none name"><code>snp.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>place_mutations()</code> generates a SNP data.frame by randomly placing a fixed
number of mutations on a given genealogy.
It means that all the sites are perfectly linked with each other.</p>
<p><code>make_snp_chromosome()</code> simulates a SNP data.frame on a chromosome using
<code>place_mutations()</code> with recombination.
<code>make_snp()</code> calls it for each chromosome in parallel.</p>
<p>Use <code>RNGkind("L'Ecuyer-CMRG")</code> and <code><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed()</a></code> to get reproducible results.
The number of CPU cores used can be configured via <code>mc.cores</code> option.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">make_snp</span><span class="op">(</span><span class="va">samples</span>, <span class="va">ss</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">make_snp_chromosome</span><span class="op">(</span><span class="va">genealogy</span>, <span class="va">segsites</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">place_mutations</span><span class="op">(</span><span class="va">genealogy</span>, <span class="va">segsites</span>, v_sampled <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-samples">samples<a class="anchor" aria-label="anchor" href="#arg-samples"></a></dt>
<dd><p>A data.frame: <code>sample_family</code> or its tranfromation by <code><a href="genealogy.html">gather_segments()</a></code>.
Using the latter will improve performance when the function is called many times,
e.g., in <code>make_snp()</code>.</p></dd>


<dt id="arg-ss">ss<a class="anchor" aria-label="anchor" href="#arg-ss"></a></dt>
<dd><p>A sequence of <code>segsites</code>;
its length is the number of chromosome;
each element is the number of segsites on a chromosome.
If a named vector is given, the output is also named.</p></dd>


<dt id="arg-genealogy">genealogy<a class="anchor" aria-label="anchor" href="#arg-genealogy"></a></dt>
<dd><p>An output from <code><a href="genealogy.html">make_gene_genealogy()</a></code>.</p></dd>


<dt id="arg-segsites">segsites<a class="anchor" aria-label="anchor" href="#arg-segsites"></a></dt>
<dd><p>The number of segregating sites on a segment.</p></dd>


<dt id="arg-v-sampled">v_sampled<a class="anchor" aria-label="anchor" href="#arg-v-sampled"></a></dt>
<dd><p>The sampled vertices. Use this to fix the output order.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>make_snp()</code> returns a list of data.frame for each chromosome.</p>
<p><code>make_snp_chromosome()</code> and <code>place_mutations()</code> returns
a data.frame with segregating sites in columns, and sample segments in rows.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>It is assumed that each node in the genealogy has one recombination event
on each chromosome. In other words, a chromosome has a fixed number of
recombination events, equal to the number of nodes in the genealogy.
Their locations are randomly assigned, and segregating sites are placed among them.
For example, a chromosome with 3 segsites on a genealogy with 10 nodes can be
illustrated like this:
<code>5' rrrSrrSrrrrSr 3'</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="vcf.html">write_vcf()</a></code> to write the output in VCF format.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">666</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">=</span> <span class="fu"><a href="tekka.html">tekka</a></span><span class="op">(</span><span class="st">"-y20 -l2 --sa 2,2 --sj 2,2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">samples</span> <span class="op">=</span> <span class="va">result</span><span class="op">$</span><span class="va">sample_family</span><span class="op">[[</span><span class="fl">1L</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">segments</span> <span class="op">=</span> <span class="fu"><a href="genealogy.html">gather_segments</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genealogy</span> <span class="op">=</span> <span class="fu"><a href="genealogy.html">make_gene_genealogy</a></span><span class="op">(</span><span class="va">segments</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">place_mutations</span><span class="op">(</span><span class="va">genealogy</span>, <span class="fl">3L</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 'data.frame':	32 obs. of  3 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ : int  1 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ : int  0 0 0 1 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ : int  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">ss</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>chr1 <span class="op">=</span> <span class="fl">3</span>, chr2 <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">make_snp</span><span class="op">(</span><span class="va">segments</span>, <span class="va">ss</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ chr1:'data.frame':	32 obs. of  3 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : int [1:32] 0 0 0 0 0 0 0 0 1 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : int [1:32] 0 0 0 0 0 1 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : int [1:32] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ chr2:'data.frame':	32 obs. of  2 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : int [1:32] 0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ : int [1:32] 0 0 0 0 0 0 0 0 0 0 ...</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Watal M. Iwasaki.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

